userspace/libsinsp/socket_handler.h

```diff
@@ -353,34 +353,67 @@ class socket_data_handler
 		}
 	}
 
-	int get_all_data()
+	int get_all_data_secure()
 	{
-		g_logger.log("Socket handler (" + m_id + ") Retrieving all data in blocking mode ...",
-					 sinsp_logger::SEV_TRACE);
-		ssize_t rec = 0;
+		int rec = 0;
+		int processed = 0;
+		int counter = 0;
+		std::vector<char> buf(1024, 0);
+		do {
+			rec = SSL_read(m_ssl_connection, &buf[0], buf.size());
+			counter++;
+			if (rec > 0)
+			{
+				processed += rec;
+			}
+			int err = SSL_get_error(m_ssl_connection, rec);
+			switch (err) {
+			case SSL_ERROR_NONE:
+				process(&buf[0], rec, false);
+				break;
+			case SSL_ERROR_ZERO_RETURN:
+				throw sinsp_exception("Socket handler (" + m_id + "): Connection closed.");
+				break;
+			case SSL_ERROR_WANT_READ:
+			case SSL_ERROR_WANT_WRITE:
+				break;
+			default:
+				g_logger.log("Socket handler (" + m_id + ") received=" + std::to_string(err) + " SSL error.",  sinsp_logger::SEV_TRACE);
+				break;
+			}
+
+			// To prevent reads from entirely stalling (like in gigantic k8s environments),
+			// give up after reading a certain size (by default, 100MB, but configurable).
+			if(processed > m_data_max_b)
+			{
+				throw sinsp_exception("Socket handler (" + m_id + "): "
+						      "read more than " + to_string(m_data_max_b / 1024 / 1024) + " MB of data from " +
+						      m_url.to_string(false) + m_path + " (" + std::to_string(processed) +
+						      " bytes, " + std::to_string(counter) + " reads). Giving up");
+			} else {
+				usleep(m_data_chunk_wait_us);
```

Comment:
Do you want to add a time-based timeout as well? I don't know the details of SSL_read() that much and how many times in a row you might get a SSL_ERROR_WANT_READ response.

Reply:
You mean something like:
`if(processed > m_data_max_b || time_spent > m_data_timeout)` ?

Reply:
If you can point me to docs that show that it's very unlikely that a loop like this can get stuck with repeated return values of SSL_ERROR_WANT_READ, that's good enough for me.

Reply:
It seems like you are right, in bold below:

> The operation did not complete; the same TLS/SSL I/O function should be called again later. If, by then, the underlying BIO has data available for reading (if the result code is SSL_ERROR_WANT_READ) or allows writing data (SSL_ERROR_WANT_WRITE), then some TLS/SSL protocol progress will take place, i.e. at least part of an TLS/SSL record will be read or written. Note that the retry may again lead to a SSL_ERROR_WANT_READ or SSL_ERROR_WANT_WRITE condition. **There is no fixed upper limit for the number of iterations that may be necessary until progress becomes visible at application protocol level.**

Should we give a fair limit to it? eg: at maximum 10 SSL_ERROR_WANT_ errors?

Reply:
I double checked some projects using openssl; all seem to use `SSL_read` same way as us.  
For example, chromium project implements a busy loop, see this [comment](https://chromium.googlesource.com/chromium/src/+/a257bd3604ae295b606b039925c56c63d62cf7b7/net/socket/ssl_client_socket_openssl.cc#1361):  note that `ERR_IO_PENDING` is a chromium internal remapping for `SSL_ERROR_WANT_READ/SSL_ERROR_WANT_WRITE` (https://chromium.googlesource.com/chromium/src/net/+/refs/heads/main/ssl/openssl_ssl_util.cc#174).

Another example is cURL: https://github.com/curl/curl/blob/60738f398cdd66312bce6ce92a87f19e71feacf4/lib/vtls/mesalink.c#L451

Although there is no fixed upper limit, it seems to me that all implementation just busy loop because at some point SSL_read() will return either the data or a different error.
Basically, those errors mean that there is still data pending and the crypto layer cannot temporarily decrypt data.

Moreover, adding a timeout would not be feasible as we cannot know in advance how long to wait for data before giving up.

Given all of the above, i think the timeout is not needed and this should be the correct way of dealing with those ssl errors.

Reply:
Ok, thanks for doing the research!